#47 parse format section

meretp opened the pull request:
This PR adds functionality to parse the new format section in the markdown files. The parsed data will also be written when generating the markdown files but I didn't know where to put the parsed information in the generated `model.ttl`. With this PR the CI in the spdx-3-model should be fixed. 

part of #47

maxhbr reviewed (commented):
some suggestions, maybe it is good to get them in.

meretp commented on spec_parser/utils.py:
```diff
@@ -330,6 +323,8 @@ def __init__(
         self._extract_metadata(metadata)
         self._extract_properties(props)
+        self.format_pattern = format_pattern
+        self.ext_props = ext_props
```
Yes, as the comment below indicates `ext_props` (and now also `format_pattern`) are silenty ignored. I could add some logic to at least write the `ext_props` to the generated markdown files similar to the fromat pattern. We still need to find out how to insert the information about external property restrictions and the new format pattern into the generated `model.ttl`

maxhbr replied:
ah, understood. Might be good to log a warning and/or to put it into the ttl as comments

meretp commented on spec_parser/parser.py:
```diff
@@ -69,6 +70,7 @@ class MDLexer(Lexer):
     DESCRIPTION = r"((?<=\n)|^)\#{2}\s+Description(?:(?!\n)\s)*(\n+|$)"
     METADATA = r"((?<=\n)|^)\#{2}\s+Metadata(?:(?!\n)\s)*(\n+|$)"
     PROPERTIES = r"((?<=\n)|^)\#{2}\s+Properties(?:(?!\n)\s)*(\n+|$)"
+    FORMAT = r"((?<=\n)|^)\#{2}\s+Format(?:(?!\n)\s)*(\n+|$)"
```
Extracted a helper method for this, only the regex for `LICENSE` is a bit different and therefore not using the new method.

meretp commented on spec_parser/utils.py:
```diff
@@ -303,20 +303,13 @@ class SpecClass(SpecBase):
-        ext_props: dict,
-        license_name: str
-    ):
+    def __init__(self, spec: Spec, namespace_name: str, name: str, summary: str, description: str, metadata: dict,
```
nope, must have been some auto-format, changed it

meretp commented on spec_parser/parser.py:
```diff
@@ -268,6 +272,27 @@ def maybe_ext_properties(self, p):
+        format_regex = re.split(r"pattern:\s", ulista, 1)[1]
+
+        return {"pattern": format_regex}
     @_("para para_line", "empty")
```
newline inserted

maxhbr commented on spec_parser/utils.py:
```diff
@@ -330,7 +333,18 @@ def __init__(
+        if ext_props:
+            self.logger.warning("External property restrictions aren't yet handled properly, they are added to the "
+                                "description of the class.")
+            self.description += "\nExternal property restriction:"
```
This probably should render the `ext_props`?

meretp replied:
added

maxhbr commented on spec_parser/utils.py:
```diff
@@ -330,7 +333,18 @@ def __init__(
+                                "description of the class.")
+            self.description += f"\nFormat pattern: {self.format_pattern['pattern']}"
+
+    # TODO: handle ext_props in some way -- for now, silently ignored
```
there should also be a symmetric comment for the format_pattern?

meretp replied:
added

maxhbr reviewed (approved):
LGTM

zvr commented on spec_parser/parser.py:
```diff
@@ -268,6 +274,28 @@ def maybe_ext_properties(self, p):
+        ulista = re.split(r"[-*+]", ulista, 1)[-1].strip()
+
+        # strip the pattern key to get only the specified regex
+        format_regex = re.split(r"pattern:\s", ulista, 1)[1]
```
Please don't add such logic in the parsing code.

Exactly like metadata section, the parsing should return a list of attribute value pairs (specified in something like `valid_metadata_keys`) and then the caller should save the data they are interested in (in this case `pattern`)

meretp replied:
Okay, I thought the "Format" section does only have this one value for "pattern". But of course if we implement this more generic we are more flexible to changes in the future. I think this results in exactly the same code as for metadata except the check for valid values. So I would reuse the methods `metadata_list` and `metadata_line` and rename them to something more generic like `list_of_attribute_value_pairs`. What would be the best way to check if the parsed section is `metadata` or `format`? My attempt would be to use `p[-3]` but this also seems a bit hacky, do you have a better idea?

zvr commented:
Actually, if both metadata and format sections are ulista, the code may be the same.

meretp commented:
@zvr I added another commit and rewrote the code so that the `metadata` and `format` section use the same parsing methods. Please have another look, it would be good if we can get this PR merged soon to fix the CI in the spdx-3-model to catch possible other formatting errors there.
